Gradient CEM-Related Work

frehe opened the pull request.

frehe committed:
Feature streamlined seeding (#28)

* Minor updates

* Add consistent use of rng's in:
- all controllers
- CartPole environment

* Consistent rng instantiation in all controllers

* #23 Minor fixes

* #23: Add info origin to log message

* Replace Python `exec` statements where possible #22

* #22: Single-line, 2 imports from string

* - Translate rng changes to new controllers
- Remove some unused imports

* Apply #22 to the new controllers

* Add explanation on seed to config

frehe committed:
- Some additional gitignores
- More req's added
- Readme update
- Updated gym loop

frehe committed:
Merge https://github.com/SensorsINI/CartPoleSimulation into Frederik_Gradient_Cem_MT

frehe committed:
Environment-agnostic predictor import

frehe committed:
Complete previous commit

frehe committed:
Remove now-unused imports:
-> Removes dependencies on cartpole predictors

frehe committed:
Pass all controller attributes in constructor #31s
- Done for all controllers intended to be portable to other environments
- Save rollout trajectories and their costs in attributes .Q, .J

frehe committed:
Encapsulation of controllers:
- Keep an environment mock
- Call costs on an instance of the selected kind
- Remove target_position (CartPole-specific) from .step(...)

frehe committed:
Merge branch 'Frederik_Feature_Import_Predictor' of https://github.com/SensorsINI/CartPoleSimulation into Frederik_Gradient_Cem_MT

frehe committed:
Some polishing

frehe committed:
Use relative imports in all externally usable controllers

frehe committed:
Update ref to its own subrepo

frehe committed:
- Update gym imlementation
- Remove relative imports

frehe committed:
Move controller template to module init

frehe committed:
Go back to correct shifting for next iteration

frehe committed:
Reference to SI_Toolkit_ASF (not global)

frehe committed:
Update ref

frehe committed:
Update ref

frehe committed:
- Rename global ASF to be installable as SI_Toolkit_ASF

frehe committed:
Add GLOBALLY_DISABLE_COMPILATION flag to package

frehe committed:
Update ref

frehe committed:
- Allow import from config even if called from parent git module
- CartPole Gym Env improvements

frehe committed:
Update ref

frehe committed:
- Controllers clip output to action space dimension
- Add mock action space to CartPole instance
    - Controllers now get action clipping limits from environment, instead of assuming them
- Add optimization termination criterion to adamresamp2 controller

frehe committed:
Minor beautification

frehe committed:
Prepare for rendering of new step API

frehe committed:
Save stored variabled more explicitly

frehe committed:
Log rollout trajectories in some controllers

frehe committed:
- Some smaller updates to existing controllers
- Add Bharadhwaj controller (different to cem naive grad)

frehe committed:
Move general controllers to Control_Toolkit

frehe committed:
Add 2 more tests

frehe committed:
Check all controllers

frehe committed:
Update ref

frehe committed:
Copy config files to ASF folder

frehe committed:
Update ref

frehe committed:
Move a utils function to control toolkit

frehe committed:
Get controller

frehe committed:
Correct import

frehe committed:
Add missing import

frehe committed:
Consistent use of dt

frehe committed:
Merge branch 'Frederik_Gradient_Cem_MT' of github.com:SensorsINI/CartPoleSimulation into Frederik_Gradient_Cem_MT

frehe committed:
No seed as NoneType instead of "None" string

frehe committed:
Use predictor names instead of types everywhere

frehe committed:
Merge from master

frehe committed:
- Run MPPI TF on physical
- Add target equilibrium variable
- Use the right config in controller_mppi

frehe committed:
Merge master into feature branch

frehe committed:
Remove path prefix from get_available_controller_names

frehe committed:
Update config

frehe committed:
Rename to predictor_name

marcinpaluch1994 commented on .SI_Toolkit_ASF/GlobalPackage/setup.py:
```diff
@@ -37,7 +37,7 @@
 
 setuptools.setup(
-    name="SI_Toolkit_ASF_global",
+    name="SI_Toolkit_ASF",
```
I would suggest deleting the file.

marcinpaluch1994 commented on CartPole/cartpole_model_tf.py:
```diff
@@ -25,9 +29,9 @@
 
-import numpy as np
-from numpy.random import SFC64, Generator
-import tensorflow as tf
+
+rng = create_rng(__name__, config["cartpole"]["seed"])
+
 
```
TODO: Double checking reproducibility & randomness

frehe replied:
Can reproduce runs when fixing all of:
* controller seed
* cartpole seed
* data generator seed

If any one is different or `null`, will change the result

marcinpaluch1994 commented on CartPole/cartpole_tf.py:
```diff
@@ -1,14 +1,20 @@
+                                      ANGLED_IDX, POSITION_IDX, POSITIOND_IDX,
+                                      STATE_INDICES)
+
+config = yaml.load(open(os.path.join(os.path.dirname(__file__), "..", "config.yml"), "r"), Loader=yaml.FullLoader)
```
I suppose we solve this with try-except?

marcinpaluch1994 commented on CartPole/cartpole_tf.py:
```diff
@@ -82,10 +90,9 @@ def Q2u_tf(Q):
-                                                                                          dtype=tf.float32) + tf.convert_to_tensor(
-        controlBias, dtype=tf.float32)
+        Q
+        + tf.convert_to_tensor(controlDisturbance * rng.standard_normal(size=Q.shape), dtype=tf.float32)
```
Won't this be float64 multiplication?

frehe replied:
I thought not. But it can be, if we want to specify it so. The controllers I've worked with compute on `float32` tensors and arrays.

What would be the advantage of `float64` over `float32`?

frehe committed:
- Delete unused ASF Global package
- Move cost weights to cartpole config

marcinpaluch1994 commented on others/Controllers-out-of-use/controller_pid.py:
```diff
@@ -3,16 +3,17 @@
 from CartPole.cartpole_model import u_max, s0
 from CartPole.cartpole_jacobian import cartpole_jacobian
 
 import yaml
```
Check!

marcinpaluch1994 commented on others/cost_functions/default.py:
```diff
@@ -1,84 +1,99 @@
-#load constants from config file
-config = yaml.load(open("config.yml", "r"), Loader=yaml.FullLoader)
+# load constants from config file
+config = yaml.load(open(os.path.join(os.path.dirname(__file__), "..", "..", "config.yml"), "r"), Loader=yaml.FullLoader)
```
Check!

marcinpaluch1994 commented on CartPole/cartpole_model_tf.py:
```diff
@@ -1,16 +1,20 @@
+                                      ANGLED_IDX, POSITION_IDX, POSITIOND_IDX,
+                                      create_cartpole_state)
+
+config = yaml.load(open(os.path.join(os.path.dirname(__file__), "..", "config.yml"), "r"), Loader=yaml.FullLoader)
```
Check

frehe replied:
Created a new helper `load_config(filename: str)` in `globals_and_utils` to perform the try/except config import where necessary. It has the advantage that if we want to later change the import precedence or places where to look for configs, then the change needs to be made only in one location

marcinpaluch1994 commented on CartPole/cartpole_model.py:
```diff
@@ -1,16 +1,22 @@
+                                      ANGLED_IDX, POSITION_IDX, POSITIOND_IDX,
+                                      create_cartpole_state)
+
+config = yaml.load(open(os.path.join(os.path.dirname(__file__), "..", "config.yml"), "r"), Loader=yaml.FullLoader)
```
Check

marcinpaluch1994 replied:
Maybe this config loading is also worth of a helper function as it comes so often... Not sure.

marcinpaluch1994 commented on GymlikeCartPole/CartPoleEnv_LTC.py:
```diff
@@ -22,20 +24,20 @@
 
-
-config = yaml.load(open("GymlikeCartPole/config_gym.yml", "r"), Loader=yaml.FullLoader)
+config = yaml.load(open(os.path.join(os.path.dirname(__file__), "config_gym.yml"), "r"), Loader=yaml.FullLoader)
```
Check

marcinpaluch1994 commented on others/Diego_semester_project/run_data_generator_legacy.py:
```diff
@@ -23,25 +20,18 @@
 config_CartPole = yaml.load(open('../../config.yml'), Loader=yaml.FullLoader)
```
Check

marcinpaluch1994 commented on GymlikeCartPole/play_gym.py:
```diff
@@ -1,18 +1,31 @@
+controller_full_name = f"controller-{controller_name}".replace("-", "_")
+Controller = import_controller_by_name(controller_full_name)
+
+config = yaml.load(open(os.path.join(os.path.dirname(__file__), "..", "config.yml"), "r"), Loader=yaml.FullLoader)
```
check

marcinpaluch1994 commented on others/Controllers-out-of-use/controller_custom_mpc_scipy.py:
```diff
@@ -1,25 +1,24 @@
+from Control_Toolkit_ASF.Controllers.controller_lqr import controller_lqr
+from others.globals_and_utils import create_rng
+from SI_Toolkit.Predictors.predictor_ODE import predictor_ODE
+
```
check

marcinpaluch1994 commented on others/Controllers-out-of-use/controller_custom_mpc_scipy.py:
```diff
@@ -1,25 +1,24 @@
+from others.globals_and_utils import create_rng
+from SI_Toolkit.Predictors.predictor_ODE import predictor_ODE
+
+config = yaml.load(open(os.path.join(os.path.dirname(__file__), "..", "..", "config.yml"), "r"), Loader=yaml.FullLoader)
```
Check

marcinpaluch1994 commented on others/Controllers-out-of-use/controller_mpc_opti.py:
```diff
@@ -14,10 +15,10 @@
 
 import yaml
-config = yaml.load(open("config.yml", "r"), Loader=yaml.FullLoader)
+config = yaml.load(open(os.path.join(os.path.dirname(__file__), "..", "..", "config.yml"), "r"), Loader=yaml.FullLoader)
```
check

marcinpaluch1994 commented on others/cost_functions/quadratic_boundary.py:
```diff
@@ -1,13 +1,22 @@
-#load constants from config file
-config = yaml.load(open("config.yml", "r"), Loader=yaml.FullLoader)
+# load constants from config file
+config = yaml.load(open(os.path.join(os.path.dirname(__file__), "..", "..", "config.yml"), "r"), Loader=yaml.FullLoader)
```
check

marcinpaluch1994 commented on others/cost_functions/quadratic_boundary_nonconvex.py:
```diff
@@ -1,14 +1,23 @@
-#load constants from config file
-config = yaml.load(open("config.yml", "r"), Loader=yaml.FullLoader)
+# load constants from config file
+config = yaml.load(open(os.path.join(os.path.dirname(__file__), "..", "..", "config.yml"), "r"), Loader=yaml.FullLoader)
```
check

marcinpaluch1994 reviewed (approved):
Read till the end, no more big questions or spotted strange things.

frehe committed:
Update refs

frehe committed:
Move SI_Toolkit ASF to folder instead of using installed package

frehe committed:
- Update set_controller
- Target position now tf Variable if env is tf
- Make sure all controllers work

marcinpaluch1994 commented:
That is just super cool, all what you are doing!

frehe committed:
- Replace config loading by try/except everywhere
    - Introduce precedence without fallback

frehe committed:
Some renaming of outdated stuff

frehe commented:
> That is just super cool, all what you are doing!

Thank you Marcin, I appreciate that
